packages/flutter/lib/src/widgets/two_dimensional_viewport.dart

```diff
@@ -0,0 +1,1216 @@
+
+  /// Represents the extent in both dimensions of the child that is actually
+  /// visible.
+  ///
+  /// For example, if a child [RenderBox] had a height of 100 pixels, and a
+  /// width of 100 pixels, but was scrolled to positions such that only 50
+  /// pixels of both width and height were visible, the paintExtent would be
+  /// represented as `Size(50.0, 50.0)`.
+  ///
+  /// This is set in the [RenderTwoDimensionalViewport.updateChildPaintData].
+  late Size paintExtent;
+
+  /// The position of the child relative to the bounds and [AxisDirection] of
+  /// the viewport.
+  ///
+  /// This is the distance from the top left visible corner of the parent to the
+  /// top left visible corner of the child. When the [AxisDirection]s are
+  /// [AxisDirection.down] or [AxisDirection.right], this value is the same as
+  /// the [layoutOffset]. This value deviates when scrolling in the reverse
+  /// directions of [AxisDirection.up] and [AxisDirection.left] to reposition
+  /// the children correctly.
+  ///
+  /// This is set in the [RenderTwoDimensionalViewport.updateChildPaintData].
+  ///
+  /// If overriding [RenderTwoDimensionalViewport.paint], use this value to
+  /// position the children instead of [layoutOffset].
+  late Offset paintOffset;
+
+  @override
+  String toString() => 'vicinity=$vicinity; layoutOffset=$layoutOffset; '
+      '${isVisible ? 'visible; ': ''}paintOffset=$paintOffset; paintExtent=$paintExtent';
+}
+
+/// A base class for viewing render objects that scroll in two dimensions.
+///
+/// The viewport listens to two [ViewportOffset]s, which determines the
+/// visible content.
+///
+/// Subclasses must implement [layoutChildSequence], calling on
+/// [buildOrObtainChildFor] to manage the children of the viewport.
+///
+/// Subclasses should not override [performLayout], as it handles housekeeping
+/// on either side of the call to [layoutChildSequence].
+// TODO(Piinks): Two follow up changes:
+//  - Keep alive <BUG>
+//  - ensureVisible <BUG>
+abstract class RenderTwoDimensionalViewport extends RenderBox implements RenderAbstractViewport {
+  /// Initializes fields for subclasses.
+  ///
+  /// The [cacheExtent], if null, defaults to [RenderAbstractViewport.defaultCacheExtent].
+  RenderTwoDimensionalViewport({
+    required ViewportOffset horizontalOffset,
+    required AxisDirection horizontalAxisDirection,
+    required ViewportOffset verticalOffset,
+    required AxisDirection verticalAxisDirection,
+    required TwoDimensionalChildDelegate delegate,
+    required Axis mainAxis,
+    required RenderObjectElement childManager,
+    double? cacheExtent,
+    Clip clipBehavior = Clip.hardEdge,
+  }) : assert(verticalAxisDirection == AxisDirection.down || verticalAxisDirection == AxisDirection.up),
+       assert(horizontalAxisDirection == AxisDirection.left || horizontalAxisDirection == AxisDirection.right),
+       _childManager = childManager as _TwoDimensionalChildManager,
```

Comment:
Yeah, this is something we discussed in the design doc, where TwoDimensionalChildManager was public, but not for any purpose other than subclasses needing to cast it when creating the render object. I thought this was what you suggested in order to avoid that cast and make it private, but I wasn't sure.

Reply:
I remember some discussion around that, but I can't find the context in the design doc anymore. I apologize if my comment steered you down a wrong path. From the current implementation it looks like TwoDimensionalChildManager should be a public class to allow proper composability of the RenderTwoDimensionalViewport. It looks like TwoDimensionalChildManager has a sensible interface, so making it public shouldn't cause any harm?

Reply:
>  I apologize if my comment steered you down a wrong path.

No need, and not at all. I wasn't sure if I had misapplied your suggestion and done something different thinking it was what you meant. :)

I found the comment, but the sample code in the doc has changed: https://docs.google.com/document/d/1C2hAq-gMAIx4Cbym7EMhVAcPZ2hmNxlgijpYmgTpmUU/edit?disco=AAAAp7OGka0

I will make this public again and see what it looks like in the test implementation.

Reply:
At one point I actually had a comment here that said

```dart
/// TODO(Piinks): I am not sure if this is what @goderbauer meant, double check :)
```